SwiftCompilerSources/Sources/Optimizer/PassManager/Context.swift

```diff
@@ -429,6 +429,33 @@ extension Type {
 //                          Builder initialization
 //===----------------------------------------------------------------------===//
 
+extension Instruction {
+  /// Returns self, unless self is a meta instruction, in which case the next
```

Comment:
We need a good definition of the term "meta instruction" - and also a better name. The term is not used anywhere else in SIL (I just figured out now that it's borrowed from LLVM's machine instructions).
In `SILInstruction::isMetaInstruction()` we define it as: "Every instruction that doesn't generate code should be in this list."
But this is not even true for `alloc_stack`: if the allocated type is generic, `alloc_stack` can generate code (metadata instantiation). On the other hand, there are a lot of instructions which really don't generate code, like `dealloc_stack`, `begin_borrow`, etc. and are not considered as "meta instructions".
So, what makes `alloc_stack` special in terms of debug info?

Reply:
> We need a good definition of the term "meta instruction" - and also a better name.

Yes we copied this term of art from MIR: https://www.llvm.org/doxygen/classllvm_1_1MachineInstr.html#aeffeb27bd92437aa2fd7b7567b01d078

> Return true if this instruction doesn't produce any output in the form of executable instructions.

I think the definition we really mean in SIL is: Should a new SIL instruction inserted after this instruction inherit this instruction's SILLocation and scope?

Since alloc_stack instructions are often hoisted to the beginning of a function, they can have a scope and location that is out-of-order. You are correct that alloc_stack instruction generate code in the form of an LLVM alloca but these instructions are part of the function prologue and don't have any scope or location.

In modern SIL alloc_stack instructions can actually have two SIL location/scope pairs, one for the instruction, and one for the variable. We could side-step the problem by setting the location of the alloc_stack instruction to a compiler-generated (or prologue?) location and relying on the variable carrying its own scope and location. Then we don't need an IsMetaInstruction helper.

Reply:
Writing this in the doc comment:
"A meta instruction is an instruction whose location is not interesting as it is impossible to set a breakpoint on it. That could be because the instruction does not generate code (such as `debug_value`), or because the generated code would be in the prologue (`alloc_stack`). Other instructions should not inherit this instruction's location."

We could add other instructions to the list, but we would have to do it on both the C++ and the Swift side, so I think it should be in a separate PR.